src/databases/SW4/avtsw4imgFileFormat.C

```diff
@@ -0,0 +1,747 @@
+
+    if (m_CartGrid)
+    {
+        mmd->name = "Cartesian grid";
+    }
+    else
+    {
+        mmd->name = "Curvilinear grid";
+    }
+
+    mmd->spatialDimension = 3;
+    mmd->topologicalDimension = 3;
+
+    mmd->numBlocks = m_nblocks;
+        
+    //   mmd->hasUnits = true;
+    //   mmd->units = "meters";
+
+    if (m_CartGrid)
+    {
+        mmd->meshType = AVT_RECTILINEAR_MESH;
+    }
+    else
+    {
+        mmd->meshType = AVT_CURVILINEAR_MESH;
+    }
+
+    debug4 << "SW4> " << "before b-loop " << endl;
+    mmd->blockNames.resize(m_nblocks);
+    for( int b=0 ; b < m_nblocks ; b++ )
+    {
+        debug4 << "SW4> " << "Block = " << b << endl;
+        debug4 << "SW4> " << "grid size = " << m_gridsize[b] << endl;
+        debug4 << "SW4> " << "no of pts = " << m_ni[b] << " " << m_nj[b]
+               << " " << m_nk[b] << endl;
+        char buf[50];
+        snprintf(buf,50,"block=%i, h=%f", b, m_gridsize[b]);
+        mmd->blockNames[b] = buf;
+    }
+    md->Add(mmd);
+    debug4 << "SW4> " << "after b-loop " << endl;
+
+    // Scalar variable information
+    avtScalarMetaData *smd1 = new avtScalarMetaData;
+    smd1->name = m_mode;
+    // copy the mesh name from above
+    smd1->meshName = mmd->name;
+    smd1->centering = AVT_NODECENT;
+    smd1->hasUnits = false;
+    md->Add(smd1);
+    debug4 << "SW4> " << "Populatedatabase done " << endl;
+}
+
+
+// ****************************************************************************
+//  Method: avtsw4imgFileFormat::GetMesh
+//
+//  Purpose:
+//      Gets the mesh associated with this file.  The mesh is returned as a
+//      derived type of vtkDataSet (ie vtkRectilinearGrid, vtkStructuredGrid,
+//      vtkUnstructuredGrid, etc).
+//
+//  Arguments:
+//      domain      The index of the domain.  If there are NDomains, this
+//                  value is guaranteed to be between 0 and NDomains-1,
+//                  regardless of block origin.
+//      meshname    The name of the mesh of interest.  This can be ignored if
+//                  there is only one mesh.
+//
+//  Programmer: petersson1 -- generated by xml2avt
+//  Creation:   Tue Nov 11 12:42:37 PDT 2014
+//
+// ****************************************************************************
+
+vtkDataSet *
+avtsw4imgFileFormat::GetMesh(int domain, const char *meshname)
+{
+    if( !m_initialized )
+        Initialize();
+
+    debug4 << "SW4> " << "entering GetMesh " << endl;
+    debug4 << "SW4> " << "meshname = '" << meshname << "'" << " domain = "
+           << domain << endl;
+
+    if (m_CartGrid || domain != m_nblocks-1)
+    {
+        debug4 << "SW4> " << "constructing a Cartesian grid, m_ni = "
+               << m_ni[domain] << " m_nj = " << m_nj[domain] << " m_nk = "
+               << m_nk[domain] << " m_gridsize = " << m_gridsize[domain]
+               << endl;
+
+        vtkFloatArray *coords[3]= {0,0,0};
+        coords[0] = vtkFloatArray::New();
+        coords[0]->SetNumberOfTuples(m_ni[domain]);
```

Comment:
Are you sure the coords will always fit within float?

Reply:
similar question with variable data, are you sure you will always be reading floats and not double?

Reply:
Updated to handle both floats and doubles.